Fixed image add link input disappears issue

faishal opened the pull request:
Fixes #5894
Same as #3822

## Description
When you begin typing anything in the image link input, it simply just closes the UI due to key events propagation, changing the states and triggering updates.

## How Has This Been Tested?
Add image block and try to add a link from the toolbar. When you start typing, input should stay open.

## Types of changes
Bug fix 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

faishal committed:
#5894:Image add link input disappear when start typing, fixed it by stopping event propagation

Signed-off-by: Faishal Saiyed <saiyedfaishal@gmail.com>

paulwilde commented:
This fixes it for me, by the way.

danielbachhuber commented:
Thanks for the pull request, @faishal 

@WordPress/gutenberg-core Do you think it's possible to add e2e tests for this?

aduth commented:
>just closes the UI due to key events propagation, changing the states and triggering updates.

Propagating to where to do what state changes?

Asking because while it occasionally has merit, I am very cautious of introducing `Event#stopPropagation` since in solving one issue it has a tendency to subtly introduce other, much more difficult to debug issues. If we need it, we should be clear at what point it fails, whether it's relevant for the component in question to have awareness of said propagation (e.g. a `components/` component shouldn't be made to tailor to the editor specifically), and ultimately to document its existence very clearly for the next maintainer.

youknowriad commented:
I rebased this but it looks like the suggestions are not showing up properly in the image link?

Commit:
Fix issues with suggestions not working in image block

The issue was that on mobile, the block toolbar has overflow auto so as to scroll horizontally. This hid the suggestions box in a weird scrollbar.

This PR tweaks that, so the suggestions don't show up on mobile at all. This is not an ideal fix, but suggestions weren't very mobile friendly in the first place.

Suggest we get this in now as a bandaid, and then look at addressing suggestions on mobile separately, for example using the same technique the block switcher uses.

jasmussen commented:
Okay, I pushed a fix to make sure the link suggestions show up when making image links. 

The issue was that on mobile, the block toolbar has overflow auto so as to scroll horizontally. This hid the suggestions box in a weird scrollbar.

I fixed so suggestions don't show up on mobile at all. This is not an ideal fix, but suggestions weren't very mobile friendly in the first place. Suggest we get this in now as a bandaid, and then look at addressing suggestions on mobile separately, for example using the same technique the block switcher uses.

Screenshots:

<img width="571" alt="screen shot 2018-04-18 at 11 57 16" src="https://user-images.githubusercontent.com/1204802/38925577-55607912-4300-11e8-96f0-763151371515.png">

<img width="767" alt="screen shot 2018-04-18 at 11 57 30" src="https://user-images.githubusercontent.com/1204802/38925582-57c5485e-4300-11e8-8bc9-564342f7819f.png">

<img width="581" alt="screen shot 2018-04-18 at 11 57 41" src="https://user-images.githubusercontent.com/1204802/38925583-5987d972-4300-11e8-83b0-875608648605.png">

Commit:
Fix issue with caption links

jasmussen commented:
Also fixed an issue with link suggestions in captions:

<img width="463" alt="screen shot 2018-04-18 at 12 32 10" src="https://user-images.githubusercontent.com/1204802/38926993-ac0bfddc-4304-11e8-9666-b6c84528b8c0.png">

youknowriad committed:
Ignore typing in block toolbars

youknowriad commented:
I also updated the way we fix the event propagation from closing the toolbar. This was happening because the `ObserveTyping` component was observing typing in link modals as well. I updated the PR to avoid observing typing in block toolbars. The classname is hardcoded but I think it's fine since it's defined in the same module.

youknowriad reviewed (approved):
LGTM 👍

youknowriad commented on editor/components/observe-typing/index.js:
```diff
@@ -120,7 +120,8 @@ class ObserveTyping extends Component {
 		// text field (e.g. arrow-ing through toolbar buttons).
-		if ( isTyping || ! isTextField( target ) ) {
+		// Ignore typing in a block toolbar
+		if ( isTyping || ! isTextField( target ) || target.closest( '.editor-block-toolbar' ) ) {
```
> This still leaves open some possibility for issues with things like link inputs in RichText though.

I think triggering isTyping in that case is the correct behavior.

-- 
For me, All inputs and richtext in block edit functions should trigger the behavior. Side UI shouldn't. All side UI is triggered is shown outside the editor content area (Popovers) aside the toolbar, so the fix made sense.

But I understand that wrapping only the `BlockEdit` component with this might also fix the issue. It would have been a good approach too.
